src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -100,28 +98,57 @@ where
         .await
         .expect("shard unavailable");
 
+    let remap_shard_write_handle_upper = write_handle.upper();
+
+    // Ensure that no collection's have advanced beyond the remap shard.
+    if let Some((id, resume_upper)) = resume_uppers
+        .iter()
+        .find(|(_, upper)| PartialOrder::less_than(remap_shard_write_handle_upper, upper))
+    {
+        mz_ore::halt!(
+            "{}'s resume upper of {:?} in advance of remap shard upper {:?}",
+            id,
+            remap_shard_write_handle_upper,
+            resume_upper
+        )
+    }
+
     let as_of = read_handle.since().clone();
+    let mut upper = as_of.clone();
 
     let mut remap_updates = vec![];
 
-    let mut subscription = read_handle
-        .subscribe(as_of.clone())
-        .await
-        .expect("always valid to read at since");
+    // If we have written any data to the remap handle, advance the upper as far as the join of the
+    // resumer uppers.
+    //
+    // Note that if we haven't written any data to the remap handle, it's not possible to open the
```

Comment:
I see that the code before this change unconditionally created the subscribe and only gated `fetch_next`. Did something change that made the previous code break? I would expect the only diff here to be changing how we calculated the `resume_upper`

Reply:
Rewrote this code so that it looks essentially the same as the old code.

Reply:
A note that we had chatted about this previously and the issue was snapshotting; however, if we just take in the resume frontier, we can rely on the old condition of exiting early.

I believe the desire was to not rely on the early exit, but instead let the work propagate down to the timestamper (which has the same condition); however, doing that requires opening a write handle to see if the remap shard's been written to––you cannot start the subscribe at the minimum tmestamp because nothing will ever close it.